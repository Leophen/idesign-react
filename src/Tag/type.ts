export interface TagProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 标签中的图标，可自定义图标呈现
   */
  icon?: string;
  /**
   * 标签最大宽度，超出部分渐变省略。示例：'50px' / 80
   */
  maxWidth?: React.CSSProperties['maxWidth'] | number;
  /**
   * 标签尺寸
   * @default medium
   */
  size?: 'small' | 'medium' | 'large';
  /**
   * 标签主题
   * @default dark
   */
  theme?: 'light' | 'dark';
  /**
   * 标签类型，用于描述组件不同的应用场景
   * @default default
   */
  type?: 'default' | 'primary' | 'warning' | 'error' | 'success';
  /**
   * 点击标签触发事件
   */
  onClick?: (e: React.MouseEvent<HTMLDivElement>) => void;
  /**
   * 添加标签触发事件，设置后会自动带上添加按钮
   */
  onAdd?: (e: React.MouseEvent<HTMLDivElement>) => void;
  /**
   * 删除标签触发事件，设置后会自动带上删除按钮
   */
  onClose?: (e: React.MouseEvent<HTMLDivElement>) => void;
}
