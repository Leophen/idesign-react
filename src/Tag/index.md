---
nav:
  title: 组件
  path: /components
group:
  title: 数据展示组件
  order: 4
order: 1
---

# Tag 标签

标签常用于标记、分类和选择

## 基本用法

标签组件可直接展示包裹的内容，也可通过 `<a>` 或 [React Router](https://reactrouter.com/) 的 `<Link>` 来嵌套标签组件作为超链接标签使用。

```tsx
import React from 'react';
import { Tag } from 'idesign-react';
import { Link } from 'react-router-dom';

const App = () => {
  return (
    <div className="idesign-demo-block-row">
      <Tag>普通标签</Tag>

      <a href="https://idesign-react.vercel.app/" target="_blank" rel="noreferrer">
        <Tag>超链接标签一</Tag>
      </a>

      <Link to="/components/button">
        <Tag>超链接标签二</Tag>
      </Link>
    </div>
  );
};

export default App;
```

## 不同样式的标签

Tag 组件提供了五种不同的颜色类型：`default`（默认）、`primary`、`success`、`warning` 和 `error`。

除此之外，可以通过 `theme` 属性来改变主题：`light`（默认）、`dark`。

```tsx
import React from 'react';
import { Tag } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Tag type="default">标签一</Tag>
        <Tag type="primary">标签二</Tag>
        <Tag type="success">标签三</Tag>
        <Tag type="warning">标签四</Tag>
        <Tag type="error">标签五</Tag>
      </div>
      <div className="idesign-demo-block-row">
        <Tag type="default" theme="dark">
          标签一
        </Tag>
        <Tag type="primary" theme="dark">
          标签二
        </Tag>
        <Tag type="success" theme="dark">
          标签三
        </Tag>
        <Tag type="warning" theme="dark">
          标签四
        </Tag>
        <Tag type="error" theme="dark">
          标签五
        </Tag>
      </div>
    </div>
  );
};

export default App;
```

## 带图标的标签

Tag 组件提供了 `icon` 属性来嵌入内置图标（[点击查看内置图标集合](./icon#全部图标)），也可以通过插槽的方式嵌入自定义图标。

```tsx
import React from 'react';
import { Tag } from 'idesign-react';
import { Icon } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-row">
      <Tag icon="TagPrice">标签一</Tag>
      <Tag>
        <Icon name="TagCollection" size={14} />
        标签二
      </Tag>
    </div>
  );
};

export default App;
```

## 带添加和删除按钮的标签

当添加 `onAdd` 添加标签事件或 `onClose` 关闭标签事件时，会自动带上添加或删除按钮。

```tsx
import React from 'react';
import { Tag } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Tag onAdd={() => alert('add')}>标签一</Tag>
        <Tag icon="TagPrice" onAdd={() => alert('add')}>
          标签二
        </Tag>
        <Tag onClose={() => alert('close')}>标签三</Tag>
        <Tag icon="TagPrice" onClose={() => alert('close')}>
          标签四
        </Tag>
      </div>
      <div className="idesign-demo-block-row">
        <Tag type="primary" onAdd={() => alert('add')}>
          标签一
        </Tag>
        <Tag type="primary" icon="TagPrice" onAdd={() => alert('add')}>
          标签二
        </Tag>
        <Tag type="primary" onClose={() => alert('close')}>
          标签三
        </Tag>
        <Tag type="primary" icon="TagPrice" onClose={() => alert('close')}>
          标签四
        </Tag>
      </div>
      <div className="idesign-demo-block-row">
        <Tag type="primary" theme="dark" onAdd={() => alert('add')}>
          标签一
        </Tag>
        <Tag type="primary" theme="dark" icon="TagPrice" onAdd={() => alert('add')}>
          标签二
        </Tag>
        <Tag type="primary" theme="dark" onClose={() => alert('close')}>
          标签三
        </Tag>
        <Tag type="primary" theme="dark" icon="TagPrice" onClose={() => alert('close')}>
          标签四
        </Tag>
      </div>
    </div>
  );
};

export default App;
```

## 不同尺寸的标签

Tag 组件提供了三种不同的尺寸：`small`、`medium`（默认）和 `large`。

```tsx
import React from 'react';
import { Tag } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Tag size="small">标签一</Tag>
        <Tag size="medium">标签二</Tag>
        <Tag size="large">标签三</Tag>
      </div>
      <div className="idesign-demo-block-row">
        <Tag size="small" icon="TagPrice">
          标签一
        </Tag>
        <Tag size="medium" icon="TagPrice">
          标签二
        </Tag>
        <Tag size="large" icon="TagPrice">
          标签三
        </Tag>
      </div>
      <div className="idesign-demo-block-row">
        <Tag size="small" onAdd={() => alert('add')}>
          标签一
        </Tag>
        <Tag size="medium" onAdd={() => alert('add')}>
          标签二
        </Tag>
        <Tag size="large" onAdd={() => alert('add')}>
          标签三
        </Tag>
      </div>
      <div className="idesign-demo-block-row">
        <Tag size="small" onClose={() => alert('close')}>
          标签一
        </Tag>
        <Tag size="medium" onClose={() => alert('close')}>
          标签二
        </Tag>
        <Tag size="large" onClose={() => alert('close')}>
          标签三
        </Tag>
      </div>
    </div>
  );
};

export default App;
```

## 自定义最大宽度

通过 `maxWidth` 设置最大宽度，超出部分自动省略。

```tsx
import React from 'react';
import { Tag } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Tag size="small">超长文本标签超长省略文本标签</Tag>
        <Tag size="small" maxWidth={150}>
          超长文本标签超长省略文本标签
        </Tag>
      </div>
      <div className="idesign-demo-block-row">
        <Tag>超长文本标签超长省略文本标签</Tag>
        <Tag maxWidth={150}>超长文本标签超长省略文本标签</Tag>
      </div>
      <div className="idesign-demo-block-row">
        <Tag size="large">超长文本标签超长省略文本标签</Tag>
        <Tag size="large" maxWidth={150}>
          超长文本标签超长省略文本标签
        </Tag>
      </div>
    </div>
  );
};

export default App;
```

<API />
