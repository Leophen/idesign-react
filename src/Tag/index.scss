.i-tag {
  display: inline-flex;
  flex-direction: row;
  align-items: center;
  box-sizing: border-box;
  height: 24px;
  padding: 0 8px;
  font-size: 14px;
  line-height: 1;
  white-space: nowrap;
  border: 1px solid transparent;
  border-radius: 4px;
  cursor: default;
  user-select: none;
}

a {
  text-decoration: none;
}

a .i-tag {
  cursor: pointer;
  transition: all 0.2s ease-out;
  &:hover {
    color: var(--i-primary);
    border-color: var(--i-primary);
  }
}

@mixin injectColor($color, $color_border, $color_border_opacity) {
  color: $color;
  border-color: $color_border;
  .i-icon {
    margin-right: 4px;
    color: $color !important;
  }

  &.i-tag--theme-dark {
    color: #fff;
    background-color: $color;
    border-color: $color;
    .i-icon {
      color: #fff !important;
    }
    &.i-tag--add-btn {
      &:hover {
        background-color: $color;
        opacity: 0.8;
      }
    }
  }

  &.i-tag--add-btn {
    cursor: pointer;
    transition: all 0.2s ease-out;
    &:hover {
      background-color: $color_border_opacity;
    }
  }

  .i-tag--close-btn {
    position: relative;
    width: 12px;
    height: 12px;
    margin-right: -2px;
    margin-left: 4px;
    cursor: pointer;
    .icon-Close {
      position: absolute;
      z-index: 1;
      margin: 0;
      transition: all 0.2s ease-out;
    }
    &::after {
      position: absolute;
      top: -1px;
      left: -1px;
      z-index: 0;
      width: calc(100% + 2px);
      height: calc(100% + 2px);
      border-radius: 50%;
      transition: all 0.2s ease-out;
      content: '';
    }
    &:hover {
      opacity: 0.8;
    }
  }
}

.i-tag--type-default {
  @include injectColor(var(--i-font-5), var(--i-font-5), var(--i-bg-active));
}

.i-tag--type-primary {
  @include injectColor(var(--i-primary), var(--i-primary), var(--i-primary-bg));
}

.i-tag--type-success {
  @include injectColor(var(--i-success), var(--i-success), var(--i-success-bg));
}

.i-tag--type-warning {
  @include injectColor(var(--i-warning), var(--i-warning), var(--i-warning-bg));
}

.i-tag--type-error {
  @include injectColor(var(--i-error), var(--i-error), var(--i-error-bg));
}

.i-tag--ellipsis {
  position: relative;
  overflow: hidden;
  text-overflow: ellipsis;
  &::after {
    position: absolute;
    right: 0;
    display: flex;
    align-items: center;
    justify-content: end;
    width: 24px;
    height: 100%;
    background: linear-gradient(270deg, var(--i-bg) 20%, rgba(255, 255, 255, 0) 100%);
    content: '';
  }
}

.i-tag--size-small {
  height: 20px;
  padding: 0 6px;
  font-size: 12px;
  .i-icon {
    margin-right: 2px;
  }
  &.i-tag--ellipsis {
    &::after {
      width: 16px;
    }
  }
}

.i-tag--size-large {
  height: 32px;
  padding: 0 12px;
  font-size: 16px;
  .i-icon {
    margin-right: 6px;
  }
  .i-tag--close-btn {
    width: 16px;
    height: 16px;
  }
  &.i-tag--ellipsis {
    &::after {
      width: 32px;
    }
  }
}
