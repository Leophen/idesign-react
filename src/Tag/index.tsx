import React from 'react';
import classNames from 'classnames';
import './index.scss';
import Icon from '../Icon';
import { TagProps } from './type';

const Tag: React.FC<TagProps> = (props) => {
  const {
    children = '',
    className,
    icon,
    maxWidth,
    size,
    style,
    theme,
    type = 'default',
    onAdd,
    onClick = () => { },
    onClose,
    ...restProps
  } = props;

  const handleClick = (e: React.MouseEvent<HTMLDivElement>) => {
    if (onAdd) {
      onAdd(e);
    }
    onClick(e);
  };

  return (
    <div
      className={classNames(
        'i-tag',
        size && `i-tag--size-${size}`,
        type && `i-tag--type-${type}`,
        theme && `i-tag--theme-${theme}`,
        onAdd && 'i-tag--add-btn',
        maxWidth && 'i-tag--ellipsis',
        className,
      )}
      style={{ ...(style || {}), ...{ maxWidth } }}
      onClick={handleClick}
      {...restProps}
    >
      {onAdd && <Icon name="ThePlus" size={size === 'large' ? 16 : 12} />}
      {icon && <Icon name={icon} size={size && { small: 12, medium: 14, large: 16 }[size]} />}
      {children}
      {onClose && (
        <div className="i-tag--close-btn" onClick={(e) => onClose(e)}>
          <Icon name="Close" size={size === 'large' ? 16 : 12} />
        </div>
      )}
    </div>
  );
};

Tag.displayName = 'Tag';

export default Tag;
