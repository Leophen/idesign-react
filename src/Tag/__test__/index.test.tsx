import { render, fireEvent } from '@testing-library/react';
import Tag from '../index';

describe('Tag 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(<Tag>标签</Tag>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('type', () => {
    const { container } = render(<Tag type="success">标签</Tag>);
    expect(container.firstChild).toHaveClass('i-tag--type-success');
  });

  it('theme', () => {
    const { container } = render(<Tag theme="dark">标签</Tag>);
    expect(container.firstChild).toHaveClass('i-tag--theme-dark');
  });

  it('icon', () => {
    const { container } = render(<Tag icon="TagPrice">标签</Tag>);
    expect(container.firstChild.firstChild).toHaveClass('icon-TagPrice');
  });

  it('onAdd', () => {
    const clickFn = jest.fn();
    const { container } = render(<Tag onAdd={clickFn}>标签</Tag>);
    expect(container.firstChild.firstChild).toHaveClass('icon-ThePlus');
    fireEvent.click(container.firstChild);
    expect(clickFn).toHaveBeenCalled();
  });

  it('onClose', () => {
    const clickFn = jest.fn();
    const { container } = render(<Tag onClose={clickFn}>标签</Tag>);
    expect(container.firstChild.lastChild).toHaveClass('i-tag--close-btn');
    fireEvent.click(container.firstChild.lastChild);
    expect(clickFn).toHaveBeenCalled();
  });

  it('size', () => {
    const { container } = render(<Tag size="small">标签</Tag>);
    expect(container.firstChild).toHaveClass('i-tag--size-small');
  });

  it('maxWidth', () => {
    const { container } = render(<Tag maxWidth={100}>标签</Tag>);
    expect(container.firstChild).toHaveStyle('max-width: 100px');
  });
});

