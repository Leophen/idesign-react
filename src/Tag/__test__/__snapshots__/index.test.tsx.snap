// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-tag i-tag--type-default"
  >
    标签
  </div>
</DocumentFragment>
`;
